src/main/java/com/slack/kaldb/metadata/core/KaldbMetadataStore.java

```diff
@@ -0,0 +1,136 @@
+    this.metadataSerializer = metadataSerializer;
+    this.logger = logger;
+
+    // Create the path to the store in ZK. However, since 2 different processes may create
+    // a path at the same time, ignore the exception if node already exists.
+    try {
+      metadataStore.create(storeFolder, "", true).get();
+    } catch (NodeExistsException e) {
+      // ignore exception, since node creation is idemponent.
+    }
+  }
+
+  protected String getPath(String snapshotName) {
+    return ZKPaths.makePath(storeFolder, snapshotName);
+  }
+
+  // TODO: byte arrays every where.
+  @SuppressWarnings("UnstableApiUsage")
+  public ListenableFuture<T> get(String path) {
+    String nodePath = getPath(path);
+    Function<String, T> deserialize =
+        new Function<>() {
+          @Override
+          public @Nullable T apply(@Nullable String data) {
+            T result;
+            try {
+              result = metadataSerializer.fromJsonStr(data);
+            } catch (InvalidProtocolBufferException e) {
+              final String msg =
+                  String.format(
+                      "Unable to de-serialize data %s at path %s into a protobuf message.",
+                      data, path);
+              logger.error(msg, e);
+              throw new CorruptMetadataNodeException(msg, e);
+            }
+            return result;
+          }
+        };
+
+    // TODO: Pass in the correct thread pool for this.
+    return Futures.transform(
+        metadataStore.get(nodePath), deserialize, MoreExecutors.directExecutor());
+  }
+
+  /**
+   * Fetches all the nodes under a given path. This function can be very expensive on nodes with a
+   * large number of children so use it very sparingly in the code. If working with slightly stale
+   * data is an option, use getCached function instead.
+   *
+   * <p>Also, if there is an issue fetching a node, we will return a null object. So, if
+   * completeness of data is important, please ignore the result of this call if there are nulls.
+   *
+   * <p>TODO: If returning nulls is problematic, fail the future if there are any failures when
+   * fetching the nodes.
+   *
+   * <p>TODO: In future, cap the number of parallel calls to ZK to a fixed number. While slow, this
+   * call will not overwhelm ZK.
+   */
+  @SuppressWarnings("UnstableApiUsage")
+  public ListenableFuture<List<T>> list() {
+    ListenableFuture<List<String>> children = metadataStore.getChildren(storeFolder);
+    Function<List<String>, List<T>> transformFunc =
+        new Function<>() {
+          @Override
+          public @Nullable List<T> apply(@Nullable List<String> paths) {
+            if (paths == null) return Collections.emptyList();
+
+            List<ListenableFuture<T>> getFutures = new ArrayList<>(paths.size());
+            for (String path : paths) {
+              getFutures.add(get(path));
+            }
+            ListenableFuture<List<T>> response = Futures.successfulAsList(getFutures);
```

Comment:
what happens to the child nodes that failed? 

`successfulAsList` will skip it silently - is that what we want? Do we have a test case where this fails ever?

Reply:
Yes, for now a partial result would be more valuable than a failed data.